app/helpers/Intersect.js

```diff
@@ -221,32 +72,233 @@ export function normalizeAndSortLocations(arr) {
+  );
+  return localDataPoints.slice(firstValid, localDataPoints.length);
+}
+
+/**
+ * Returns an array with <exposureWindowDays> elements.
+ * The elements have a value of -1 if there are no local
+ * data points for that day, otherwise zero.
+ *
+ * This method should be called before filling the gaps
+ * in the local points data because the filled data would
+ * generate the same result as filling day bins with zeros.
+ *
+ * @param {array} localDataPoints
+ * @param {int} exposureWindowDays
+ */
+export function initLocationBins(
+  exposureWindowDays = MAX_EXPOSURE_WINDOW_DAYS,
+  localDataPoints = [],
+) {
+  dayjs.extend(dayOfYear);
+  const todayDOY = dayjs().dayOfYear();
+  const dayBins = new Array(exposureWindowDays).fill(-1);
+
+  for (let ldp of localDataPoints) {
+    const daysAgo = todayDOY - dayjs(ldp.time).dayOfYear();
+    if (dayBins[daysAgo] === 0 || daysAgo >= exposureWindowDays) continue;
+    dayBins[daysAgo] = 0;
+  }
+  return dayBins;
+}
+
+/**
+ * Returns an array that consists all elements from <localData> array,
+ * with elements inserted between two local data points if there is a time gap
+ * larger than <gpsPeriodMS>.
+ *
+ * The number of elements inserted between two elements is one less
+ * than the number of gps periods that ocurred between the two points.
+ *
+ * @param {array} localData - array of stored gps points with gaps
+ * @param {int} gpsPeriodMS - local points sampling period
+ */
+export function fillLocationGaps(
+  localData,
+  gpsPeriodMS = MIN_LOCATION_UPDATE_MS,
+) {
+  // helper function that creates and populates an array with correct timestamps
+  const generateGapPoints = (startTime, gapSize, gpsPeriodMS) =>
+    [...new Array(gapSize)].map((_, i) => ({
+      time: startTime + (i + 1) * gpsPeriodMS,
+      hashes: [],
+    }));
+
+  const filled = [];
+  for (let i = 0; i < localData.length; i++) {
+    filled.push(localData[i]);
+
+    if (i === localData.length - 1) continue;
+    // gap size is the amount of additional gps sampling periods
+    // that can fit in the interval between two local data points
+    const interval = dayjs(localData[i + 1].time).diff(localData[i].time);
+    const gapSize = Math.round(interval / gpsPeriodMS) - 1;
+    if (gapSize > 0) {
+      // generate missing data points with correct times
+      filled.push(
+        ...generateGapPoints(localData[i].time, gapSize, gpsPeriodMS),
+      );
+    }
+  }
+
+  return filled;
+}
+
+/**
+ * Updates the `hasMatch` flag of recorded GPS data points
+ * that are in the HA's list of concern points.
+ *
+ * This method does not calculate the exposure durations,
+ * as that can be done only after all of the HA data chunks
+ * have been analyzed.
+ *
+ * @param {array} localGPSDataPoints - array of recorded GPS data points.
+ * @param {Set} concernPointHashes - set of hashed concern points issued by health authorities
+ */
+export function updateMatchFlags(localGPSDataPoints, concernPointHashes) {
```

Comment:
I can make not-mutative, but it will still return the whole list of points with the `hasMatch` attribute set to true or false. Is that behaviour acceptable?